File: t/PACSFramework/VB/CustomizingWorklistDAL/Form1.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.IO
Imports System.Threading
Imports Leadtools.Dicom
Imports Leadtools
Imports Leadtools.Medical.DataAccessLayer
Imports Leadtools.Dicom.Scp.Command
Imports Leadtools.Medical.Worklist.DataAccessLayer
Imports Leadtools.Demos

Imports VBCustomizingWorklistDAL.DataTypes

Namespace VBCustomizingWorklistDAL
   Partial Public Class Form1 : Inherits Form
      Private _databaseLayer As DatabaseLayer
      Private _source As WorklistDataSource
      Private _worklistAgent As IWorklistDataAccessAgent
      Private _iodPath As String

      Public ReadOnly Property DatabaseLayer() As DatabaseLayer
         Get
            Return _databaseLayer
         End Get
      End Property

      ''' <summary>
      ''' The main entry point for the application.
      ''' </summary>
      '<STAThread()> _
      Shared Sub Main()
         If Not Support.SetLicense() Then
            Return
         End If

         Application.EnableVisualStyles()
         Application.SetCompatibleTextRenderingDefault(False)
         Application.Run(New Form1())
      End Sub

      Public Sub New()
         InitializeComponent()

         DicomEngine.Startup()

         Messager.Caption = Me.Text

         If (Not DatabaseLayer.IsDataAccessSettingsValid()) Then
            MessageBox.Show("Database settings is not valid. Please run the PACS Database Configuration Demo.", Me.Text, MessageBoxButtons.OK, MessageBoxIcon.Error)

            Return
         End If

         _source = New WorklistDataSource()
         _databaseLayer = New DatabaseLayer(_source)
         _iodPath = Application.StartupPath

         _iodPath = Path.Combine(_iodPath, "CustomMWLIOD.xml")

         Using iodStream As Stream = MWLCFindCommand.DefualtMWLIOD
            Using customIODStream As FileStream = New FileStream(_iodPath, FileMode.Create)
               CopyTo(iodStream, customIODStream)
            End Using
         End Using

         _source.DatabaseTags.Add(New DatabaseDicomTags("Visit", "ServiceEpisodeDescription", DicomTag.ServiceEpisodeDescription))
         _source.DatabaseTags.Add(New DatabaseDicomTags("ScheduledProcedureStep", "CommentsOnTheScheduledProcedureStep", DicomTag.CommentsOnTheScheduledProcedureStep))
         _source.DatabaseTags.Add(New DatabaseDicomTags("Patient", "SmokingStatus", DicomTag.SmokingStatus))

         _worklistAgent = _databaseLayer.GetWorklistDataAgent()

         DataAccessServices.RegisterDataAccessService(Of IWorklistDataAccessAgent)(_worklistAgent)

         databaseStatus1.ConnectionString = _databaseLayer.ConnectionString
         databaseStatus1.ProviderName = _databaseLayer.Provider

         dicomTags1.SetSource(_source.DatabaseTags)
         worklistUpdate1.SetSource(_source.DatabaseTags)
         dicomQuery1.SetSource(_source.DatabaseTags, _iodPath)

         If _databaseLayer.IsDatabaseUpdated() Then
            Try
               WorklistIODUpdater.UpdateIOD(_source.DatabaseTags, _iodPath)
            Catch exception As Exception
               Messager.ShowError(Me, "Error Updating the Modality Work-list IOD document." + Environment.NewLine + exception.Message)
            End Try
         End If

         AddHandler UpdateDatabaseButton.Click, AddressOf UpdateDatabaseButton_Click
      End Sub

      Private Sub UpdateDatabaseButton_Click(ByVal sender As Object, ByVal e As EventArgs)
         Try
            If _databaseLayer.IsDatabaseUpdated() Then
               Messager.ShowInformation(Me, "Database has already been updated.")
            Else
               ' Add the tags to _source, only if they have not already been added
               Try
                  WorklistIODUpdater.UpdateIOD(_source.DatabaseTags, _iodPath)
               Catch exception As Exception
                  Messager.ShowError(Me, "Error Updating the Modality Work-list IOD document." + Environment.NewLine + exception.Message)
               End Try

               _databaseLayer.UpdateDatabase()
               Messager.ShowInformation(Me, "Database updated successfully.")
            End If

         Catch exception As Exception
            Messager.ShowError(Me, exception)
         End Try
      End Sub

      Private Shared Sub CopyTo(ByVal src As Stream, ByVal dest As Stream)
         If src Is Nothing Then
            Throw New System.ArgumentNullException("src")
         End If

         If dest Is Nothing Then
            Throw New System.ArgumentNullException("dest")
         End If

         System.Diagnostics.Debug.Assert(src.CanRead, "src.CanRead")
         System.Diagnostics.Debug.Assert(dest.CanWrite, "dest.CanWrite")

         Dim readCount As Integer = 0
         Dim buffer As Byte() = New Byte(8191) {}
         readCount = src.Read(buffer, 0, buffer.Length)
         Do While readCount <> 0
            dest.Write(buffer, 0, readCount)
            readCount = src.Read(buffer, 0, buffer.Length)
         Loop

      End Sub

   End Class
End Namespace

File: t/PACSFramework/CS/CustomizingWorklistDAL/Form1.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using CSCustomizingWorklistDAL.DataTypes;
using Leadtools.Dicom;
using Leadtools;
using CSCustomizingWorklistDAL.MyClientSession;
using System.Threading;
using Leadtools.Medical.DataAccessLayer;
using Leadtools.Dicom.Scp.Command;
using Leadtools.Medical.Worklist.DataAccessLayer;
using System.IO;
using Leadtools.Demos;

namespace CSCustomizingWorklistDAL
{
   public partial class Form1 : Form
   {
      DatabaseLayer _databaseLayer;
      WorklistDataSource       _source ;
      IWorklistDataAccessAgent _worklistAgent ;
      string                   _iodPath ;

      public DatabaseLayer DatabaseLayer
      {
         get { return _databaseLayer; }
      }
      
      public Form1 ( )
      {
         InitializeComponent ( ) ;
         
         Messager.Caption = this.Text ;
         
         if ( !DatabaseLayer.IsDataAccessSettingsValid ( ) )
         {
            MessageBox.Show ( "Database settings is not valid. Please run the PACS Database Configuration Demo.", this.Text, MessageBoxButtons.OK, MessageBoxIcon.Error ) ;
            
            return ;
         }
         
         _source        = new WorklistDataSource ( ) ;
         _databaseLayer = new DatabaseLayer ( _source ) ;
         _iodPath       = Application.StartupPath ;
         
         _iodPath = Path.Combine ( _iodPath, "CustomMWLIOD.xml" ) ;
         
         using ( Stream iodStream = MWLCFindCommand.DefualtMWLIOD )  
         {
            using ( FileStream customIODStream = new FileStream ( _iodPath, FileMode.Create ) )
            {
               CopyTo ( iodStream, customIODStream ) ;
            }
         }

         _source.DatabaseTags.Add(new DatabaseDicomTags("Visit", "ServiceEpisodeDescription", DicomTag.ServiceEpisodeDescription));
         _source.DatabaseTags.Add(new DatabaseDicomTags("ScheduledProcedureStep", "CommentsOnTheScheduledProcedureStep", DicomTag.CommentsOnTheScheduledProcedureStep));
         _source.DatabaseTags.Add(new DatabaseDicomTags("Patient", "SmokingStatus", DicomTag.SmokingStatus));
         
         _worklistAgent  = _databaseLayer.GetWorklistDataAgent ( ) ;
         
         DataAccessServices.RegisterDataAccessService <IWorklistDataAccessAgent> ( _worklistAgent ) ;
         
         databaseStatus1.ConnectionString = _databaseLayer.ConnectionString ;
         databaseStatus1.ProviderName     = _databaseLayer.Provider ;
         
         dicomTags1.SetSource      ( _source.DatabaseTags ) ;
         worklistUpdate1.SetSource ( _source.DatabaseTags ) ;
         dicomQuery1.SetSource     ( _source.DatabaseTags, _iodPath ) ;

         if (_databaseLayer.IsDatabaseUpdated())
         {
            try
            {
               WorklistIODUpdater.UpdateIOD(_source.DatabaseTags, _iodPath);
            }
            catch (Exception exception)
            {
               Messager.ShowError(this, "Error Updating the Modality Work-list IOD document.\n" + exception);
            }
         }
         
         UpdateDatabaseButton.Click += new EventHandler(UpdateDatabaseButton_Click);
      }

      void UpdateDatabaseButton_Click(object sender, EventArgs e)
      {
         try
         {
            if (_databaseLayer.IsDatabaseUpdated())
            {
               Messager.ShowInformation(this, "Database has already been updated.");
            }
            else
            // Add the tags to _source, only if they have not already been added
            {
               try
               {
                  WorklistIODUpdater.UpdateIOD(_source.DatabaseTags, _iodPath);
               }
               catch (Exception exception)
               {
                  Messager.ShowError(this, "Error Updating the Modality Work-list IOD document.\n" + exception);
               }

               _databaseLayer.UpdateDatabase();
               Messager.ShowInformation(this, "Database updated successfully.");
            }
            
         }
         catch ( Exception exception ) 
         {
            Messager.ShowError ( this, exception ) ;
         }
      }
      
      private static void CopyTo ( Stream src, Stream dest )
      {
         if (src == null)
         {
            throw new System.ArgumentNullException("src");
         }
         
         if (dest == null)
         {      
            throw new System.ArgumentNullException ( "dest" ) ;
         }

         System.Diagnostics.Debug.Assert(src.CanRead, "src.CanRead");
         System.Diagnostics.Debug.Assert(dest.CanWrite, "dest.CanWrite");

         int readCount;
         var buffer = new byte[8192];
         while ((readCount = src.Read(buffer, 0, buffer.Length)) != 0)
         {
            dest.Write(buffer, 0, readCount);
         }
      }

   }
}
